cmd/kube-spawn/create.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2017 Kinvolk GmbH
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"log"
+	"os"
+	"path"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/kinvolk/kube-spawn/pkg/bootstrap"
+	"github.com/kinvolk/kube-spawn/pkg/config"
+	"github.com/kinvolk/kube-spawn/pkg/utils"
+)
+
+var (
+	createCmd = &cobra.Command{
+		Use:   "create",
+		Short: "Generates the environment for a cluster. If you change 'kspawn.toml' this needs to be run again.",
+		Run:   runCreate,
+	}
+)
+
+func init() {
+	kubespawnCmd.AddCommand(createCmd)
+
+	// do not set defaults here
+	// intead use:
+	// pkg/config/defaults.go
+	// and call from the if uninitialized {} block below
+	//
+	createCmd.Flags().StringP("container-runtime", "r", "", "runtime to use for the spawned cluster (docker or rkt)")
+	createCmd.Flags().String("kubernetes-version", "", `version kubernetes used to initialize the cluster. Irrelevant if used wih --dev. Only accepts semantic version strings like "v1.7.5"`)
+	createCmd.Flags().Bool("dev", false, "create a cluster from a local build of Kubernetes")
+	createCmd.Flags().IntP("nodes", "n", 0, "number of nodes to spawn")
+	createCmd.Flags().StringP("image", "i", "", "base image for nodes")
+	viper.BindPFlags(createCmd.Flags())
+
+	viper.BindEnv("runtime-config.rkt.rkt-bin", "KUBE_SPAWN_RKT_BIN")
```

Comment:
maybe this is just me but I don't like the idea of having cli flags for all of these binaries. It's more or less necessary for the stage1 image but if you have installed rkt with your distros pkg manager and/or have `$GOPATH/bin` in your `PATH` you won't need to configure it anyway...

Reply:
Also we have the problem that not all of the options in the config file are possible to be set from env/cli (or make sense to be) - like the path mapping for example.

This might be solvable by having two structures, one that viper binds to and the other we handle ourselves.

Reply:
Yeah, probably user and cluster configuration should be split.